CMake: Address review comment from PR#13566

rajkan01 opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->
Addressed review comments from @rwalton-arm in the PRhttps://github.com/ARMmbed/mbed-os/pull/13566
- Replaced mbed_set_language_standard with target_compile_features.
- Renamed input source files from CMakelists.txt to CMakeLists.txt.
- Simplify the inclusion of cryptocell310 library files.
- Added more information to TODO comment.
- Removed empty input source files.

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
None.
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
None.
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->
None.
<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [x] Tests / results supplied as part of this PR
    
Manula testing:
Successful compilation of the blinky example based on CMake.

----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->
@0xc0170 @hugueskamba @rwalton-arm 
<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

0xc0170 commented on CMakeLists.txt:
```diff
@@ -109,7 +105,11 @@ add_subdirectory(cmsis/device/rtos EXCLUDE_FROM_ALL)
 #
 function(mbed_configure_app_target target)
-    mbed_set_language_standard(${target})
+    target_compile_features(${target}
```
shouldn't this be in our toolchain file (related to toolchain set up, although for an app)

rajkan01 replied:
I reverted the changes and keep the `mbed_set_language_standard` funcion and moved `target_compile_feature` inside of it, please re-review

hugueskamba commented on tools/cmake/toolchain.cmake:
```diff
@@ -48,18 +48,10 @@ set(CMAKE_CXX_COMPILER_WORKS TRUE)
-            CXX_STANDARD 14
-            CXX_STANDARD_REQUIRED YES
-            CXX_EXTENSIONS YES
+    target_compile_features(${target}
```
The idea from Rob's comment was to remove `mbed_set_language_standard()` and just use `target_compile_features()`. Not to have `mbed_set_language_standard()` as a wrapper.

0xc0170 replied:
I see, that would work. 👍

0xc0170 replied:
Making it not breaking:
1. Update all examples to not to use this function
2. remove the function from our CMake (separate PR)

rajkan01 replied:
@0xc0170  I have tried locally with removed `mbed_configure_app_target` and kept `target_compile_features` in `toolchain.cmake`, compilation fails for me, is it working for you?

hugueskamba replied:
> Making it not breaking:
> 
> 1. Update all examples to not to use this function
> 2. remove the function from our CMake (separate PR)

Applications use `mbed_configure_app_target()` which can perhaps remain as such in case we need to add more stuff in the future. `mbed_configure_app_target()` however must use `target_compile_features()` directly and not `mbed_set_language_standard()`. Therefore, there would not be a need to update examples.

hugueskamba commented on connectivity/drivers/mbedtls/FEATURE_CRYPTOCELL310/TARGET_MCU_NRF52840/CMakeLists.txt:
```diff
@@ -1,36 +1,16 @@
-_mbed_get_libcc_310_ext()
-_mbed_get_libcc_310_trng()
+if(${MBED_TOOLCHAIN} STREQUAL "GCC_ARM")
+    set(LIBCC_310_EXT TOOLCHAIN_GCC_ARM/libcc_310_ext.a)
+    set(LIBCC_310_TRNG TOOLCHAIN_GCC_ARM/libcc_310_trng.a)
+elseif(${MBED_TOOLCHAIN} STREQUAL "ARM")
+    set(LIBCC_310_EXT TOOLCHAIN_ARM/libcc_310_ext.ar)
+    set(LIBCC_310_TRNG TOOLCHAIN_ARM/libcc_310_trng.ar)
+elseif(${MBED_TOOLCHAIN} STREQUAL "IAR")
+    set(LIBCC_310_EXT TOOLCHAIN_IAR/lib_cc310_ext.a)
+    set(LIBCC_310_TRNG TOOLCHAIN_IAR/lib_cc310_trng.a)
```
```suggestion
    set(LIBCC_310_EXT ./TOOLCHAIN_GCC_ARM/libcc_310_ext.a)
    set(LIBCC_310_TRNG ./TOOLCHAIN_GCC_ARM/libcc_310_trng.a)
elseif(${MBED_TOOLCHAIN} STREQUAL "ARM")
    set(LIBCC_310_EXT ./TOOLCHAIN_ARM/libcc_310_ext.ar)
    set(LIBCC_310_TRNG ./TOOLCHAIN_ARM/libcc_310_trng.ar)
elseif(${MBED_TOOLCHAIN} STREQUAL "IAR")
    set(LIBCC_310_EXT ./TOOLCHAIN_IAR/lib_cc310_ext.a)
    set(LIBCC_310_TRNG ./TOOLCHAIN_IAR/lib_cc310_trng.a)
```

hugueskamba commented on connectivity/drivers/mbedtls/FEATURE_CRYPTOCELL310/TARGET_MCU_NRF52840/CMakeLists.txt:
```diff
@@ -40,4 +20,10 @@ target_include_directories(mbed-mbedtls-cryptocell310
+target_link_libraries(mbed-mbedtls-cryptocell310
+    INTERFACE
+        ./${LIBCC_310_EXT}
+        ./${LIBCC_310_TRNG}
```
```suggestion
        ${LIBCC_310_EXT}
        ${LIBCC_310_TRNG}
```

hugueskamba commented on connectivity/drivers/mbedtls/FEATURE_CRYPTOCELL310/binaries/CMakeLists.txt:
```diff
@@ -1,18 +1,15 @@
-    )
-endfunction()
+if(${MBED_TOOLCHAIN} STREQUAL "GCC_ARM")
+    set(LIBCC_310_CORE TOOLCHAIN_GCC_ARM/libcc_310_core.a)
+elseif(${MBED_TOOLCHAIN} STREQUAL "ARM")
+    set(LIBCC_310_CORE TOOLCHAIN_ARM/libcc_310_core.ar)
+elseif(${MBED_TOOLCHAIN} STREQUAL "IAR")
+    set(LIBCC_310_CORE TOOLCHAIN_IAR/lib_cc310_ext.a)
```
```suggestion
    set(LIBCC_310_CORE ./TOOLCHAIN_GCC_ARM/libcc_310_core.a)
elseif(${MBED_TOOLCHAIN} STREQUAL "ARM")
    set(LIBCC_310_CORE ./TOOLCHAIN_ARM/libcc_310_core.ar)
elseif(${MBED_TOOLCHAIN} STREQUAL "IAR")
    set(LIBCC_310_CORE ./TOOLCHAIN_IAR/lib_cc310_ext.a)
```

hugueskamba replied:
https://github.com/ARMmbed/mbed-os/pull/13870#discussion_r524324187

hugueskamba commented on connectivity/drivers/mbedtls/FEATURE_CRYPTOCELL310/binaries/CMakeLists.txt:
```diff
@@ -1,18 +1,15 @@
-_mbed_get_libcc_310_core()
+target_link_libraries(mbed-mbedtls-cryptocell310
+    INTERFACE
+        ${CMAKE_CURRENT_SOURCE_DIR}/${LIBCC_310_CORE}
```
```suggestion
        ./${LIBCC_310_CORE}
```

0xc0170 replied:
this is one of the exceptions, other other one is linker script as well. 

It's recommended to use for libraries the full path, this is quoted from CMake docs (https://cmake.org/cmake/help/latest/command/target_link_directories.html)

> Note This command is rarely necessary and should be avoided where there are other choices. Prefer to pass full absolute paths to libraries where possible, since this ensures the correct library will always be linked. The find_library() command provides the full path, which can generally be used directly in calls to target_link_libraries(). Situations where a library search path may be needed include:

rajkan01 committed:
CMake: Replace mbed_set_language_standard with target_compile_features and remove MBED_TARGET_LABELS variable resetting

rajkan01 committed:
CMake: Rename input source files from CMakelists.txt to CMakeLists.txt

rajkan01 committed:
CMake: Remove empty input source files

rajkan01 committed:
CMake: Add more information to TODO comment

rajkan01 committed:
CMake: Simplify inclusion of cryptocell310 library files

rwalton-arm reviewed (dismissed):
LGTM

rajkan01 committed:
Remove add subdirectory of TARGET_MCU_NRF52840 as it contains only mbed_lib.json

0xc0170 commented on connectivity/drivers/ble/FEATURE_BLE/TARGET_NORDIC/TARGET_NORDIC_CORDIO/CMakeLists.txt:
```diff
@@ -1,8 +1,6 @@
-if("MCU_NRF52840" IN_LIST MBED_TARGET_LABELS)
-    add_subdirectory(TARGET_MCU_NRF52840)
-elseif("NRF5x" IN_LIST MBED_TARGET_LABELS)
+if("NRF5x" IN_LIST MBED_TARGET_LABELS)
```
was there an intention here? The build failed to built this target with cordio sources. please review logs

0xc0170 replied:
I see, this makes sense. So why it fails now then?

rajkan01 committed:
Fix CI issue:

- Added absolute path as cmake failed to find linker file from relative path
- Added the missing MBED_CONF_CRYPTOCELL310_PRESENT
- Added the dependency library to BLE Cordio stack

hugueskamba commented on connectivity/drivers/mbedtls/FEATURE_CRYPTOCELL310/TARGET_MCU_NRF52840/CMakeLists.txt:
```diff
@@ -40,4 +20,10 @@ target_include_directories(mbed-mbedtls-cryptocell310
+
+target_link_libraries(mbed-mbedtls-cryptocell310
+    INTERFACE
+        ${CMAKE_CURRENT_SOURCE_DIR}/${LIBCC_310_EXT}
+        ${CMAKE_CURRENT_SOURCE_DIR}/${LIBCC_310_TRNG}
```
```suggestion
        ./${LIBCC_310_EXT}
        ./${LIBCC_310_TRNG}
```

hugueskamba replied:
https://github.com/ARMmbed/mbed-os/pull/13870#discussion_r524324187
